re #931 SOAP Authorization Policy

kahboom opened the pull request:
Added `SOAP Authorization Policy`.

JIRA: https://issues.jboss.org/browse/APIMAN-931
## 
#### Screenshots

<img width="1192" alt="screen shot 2016-02-15 at 4 53 26 pm" src="https://cloud.githubusercontent.com/assets/3844502/13061292/2202f868-d405-11e5-8a5e-58d7a7484cc0.png">
## 

<img width="1293" alt="screen shot 2016-02-15 at 4 53 18 pm" src="https://cloud.githubusercontent.com/assets/3844502/13061293/2440fe68-d405-11e5-80d6-e63f420b4387.png">
## 

<img width="962" alt="screen shot 2016-02-15 at 4 57 58 pm" src="https://cloud.githubusercontent.com/assets/3844502/13061767/010298aa-d408-11e5-9f19-3e39361be70d.png">
## 

<img width="977" alt="screen shot 2016-02-15 at 8 03 59 pm" src="https://cloud.githubusercontent.com/assets/3844502/13064273/58e0dfc0-d41f-11e5-923b-3adae020a408.png">

cc @EricWittmann

kahboom committed:
re #931 SOAPAuthorization Policy Plugin

EricWittmann commented on soap-authorization-policy/src/main/apiman/policyDefs/schemas/soap-authorization-PolicyDef.schema:
```diff
@@ -0,0 +1,50 @@
+      "type": "string",
+      "enum": [
+        "I want the request to pass when all matching authorization rules pass.",
+        "I want the request to pass when at least one matching authorization rules pass."
```
I suspect this is not what you want.

EricWittmann replied:
I would expect the values "all" and "any" to be possible values here?

kahboom replied:
The reason I did it this way is because of the formatting in the UI. For example, we would need to split it into three parts:
1. "I want the request to pass when "
2. ["all", "at least one"]
3. " matching authorization rules pass."

Each property is automatically displayed on a new line unless it is something like an array, but we'd want all three parts to be on the same line as it is for the Authorization Policy.

It is difficult to use CSS to ensure that all three values are on the same line because forms are not assigned unique classes or id's, but what I can do is see if we can use the grid layout instead.

kahboom replied:
Or we could also just have three lines for it, if it's not a big deal. Up to you.

EricWittmann replied:
I think you can add a 'description' property to the 'multiMatch' object.  The description can explain what the values mean.  The UI doesn't need to mirror the existing auth policy UI (which is likely not possible).  But the enum values really do need to reflect the values in the Java enum or else this will not work.

kahboom replied:
No problem, I'll update it asap.

EricWittmann commented on soap-authorization-policy/src/main/apiman/policyDefs/schemas/soap-authorization-PolicyDef.schema:
```diff
@@ -0,0 +1,50 @@
+      "type": "string",
+      "enum": [
+        "I want the request to fail if it does not match any of the authorization rules defined above.",
+        "I want the request to pass if it does not match any of the authorization rules defined above."
```
Same comment as above.  The enum values should presumably match the possible values in the java enumeration.

kahboom commented:
@EricWittmann - See below and let me know what you think. Also, let me know if there is an issue with capitalizing the first letter.

<img width="922" alt="screen shot 2016-02-16 at 2 43 42 pm" src="https://cloud.githubusercontent.com/assets/3844502/13088802/c78b892e-d4bb-11e5-9fc6-46299cf5a697.png">
<img width="943" alt="screen shot 2016-02-16 at 2 43 51 pm" src="https://cloud.githubusercontent.com/assets/3844502/13088803/c7947778-d4bb-11e5-948f-587aa3e8ffae.png">

EricWittmann commented:
Have you tried actually configuring these options in the UI and then publishing the resulting policy config to the gateway and sending a request?  I think that will answer your question.  :)

Spoiler alert - I'm guessing those options wont' work.  The UI looks good but if the values don't match the acceptable values in the java enum......

Example:
https://github.com/apiman/apiman/blob/master/gateway/engine/policies/src/main/java/io/apiman/gateway/engine/policies/config/MultipleMatchType.java

kahboom commented:
@EricWittmann - I changed it, but could not test. Will test tomorrow morning. Also, the build is failing- something about not being able to find the contextAttributes method:

```
symbol:   method contextAttribute(java.lang.String,java.util.HashSet<java.lang.String>)
location: variable request of type io.apiman.test.policies.PolicyTestRequest
```

kahboom commented:
@EricWittmann - Should be ready to go, I forgot to enable this repo in my Travis CI, which is likely why the build was failing.

EricWittmann commented:
So were you able to do a manual end-to-end test using the UI to configure the policy, then deploying it to the gateway and testing it?

msavy commented:
Is this the latest code BTW? Happy to do a review if that's helpful.

EricWittmann commented:
More eyes would be good, esp. as rachel continues her java learning.

On Thu, Feb 18, 2016 at 12:37 PM, Marc Savy notifications@github.com
wrote:

> Is this the latest code BTW? Happy to do a review if that's helpful.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/apiman/apiman-plugins/pull/55#issuecomment-185828305.

kahboom commented:
@EricWittmann - Not sure if I told you, but authorized requests are not passing through, so please don't merge this in yet while I look into it.

kahboom commented:
@EricWittmann @msavy - Apologies for the delay. The latest commit should fix the problem. Authorized requests were being denied because the `SOAPAction` header was not being passed through and I had set unmatched requests to fail in the UI. I updated the JSON schema to pass the correct value and now everything appears to be working fine. Unit tests were passing because they manually add in the `SOAPAction` header (E2E tests would/should have caught this). See below for some screenshots of my manual tests.

Given two users, `bwayne` and `ckent`:
- `bwayne` roles: `admin`, `user`
- `ckent` roles: `user`

**1. Fail Unmatched Requests:**

<img width="865" alt="screen shot 2016-02-19 at 8 58 30 am" src="https://cloud.githubusercontent.com/assets/3844502/13178189/88a7a3c8-d6eb-11e5-8e20-eed8b28ea650.png">

<img width="1284" alt="screen shot 2016-02-19 at 9 21 43 am" src="https://cloud.githubusercontent.com/assets/3844502/13178193/8b3e95c4-d6eb-11e5-85e5-ff57fd8822b3.png">

**2. Pass Unmatched Requests:**

<img width="869" alt="screen shot 2016-02-19 at 9 22 12 am" src="https://cloud.githubusercontent.com/assets/3844502/13178195/90410ca0-d6eb-11e5-847d-22f7ae1bdaff.png">

<img width="1271" alt="screen shot 2016-02-19 at 9 23 36 am" src="https://cloud.githubusercontent.com/assets/3844502/13178201/94cdb14c-d6eb-11e5-8d77-fb731edef2ef.png">

I also updated the label from `Ignored URI` to `Path`; a bug I blindly copied over from the original Authorization Policy plugin JSON schema. Finally, squashing was getting ugly, so apologies for the 2 extra commits.

Please let me know if you find anything you need for me to fix/change. Thanks.

EDIT: Squashed down to one commit, so all is good in the hood.
